### PR TITLE
[SPARK-13329] [SQL] considering output for statistics of logical plan

### PR DESCRIPTION
**Pull request by davies**
The current implementation of statistics of UnaryNode does not considering output (for example, Project may product much less columns than it's child), we should considering it to have a better guess.

We usually only join with few columns from a parquet table, the size of projected plan could be much smaller than the original parquet files. Having a better guess of size help we choose between broadcast join or sort merge join.

After this PR, I saw a few queries choose broadcast join other than sort merge join without turning spark.sql.autoBroadcastJoinThreshold for every query, ended up with about 6-8X improvements on end-to-end time.

We use `defaultSize` of DataType to estimate the size of a column, currently For DecimalType/StringType/BinaryType and UDT, we are over-estimate too much (4096 Bytes), so this PR change them to some more reasonable values. Here are the new defaultSize for them:

DecimalType:  8 or 16 bytes, based on the precision
StringType:  20 bytes
BinaryType: 100 bytes
UDF: default size of SQL type

These numbers are not perfect (hard to have a perfect number for them), but should be better than 4096.

### CODE REVIEW

**Comment by andrewor14** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 1]
```diff
@@ -316,6 +316,21 @@ abstract class UnaryNode extends LogicalPlan {
...
 
   override protected def validConstraints: Set[Expression] = child.constraints
+
+  override def statistics: Statistics = {
+    // There should be some overhead in Row object, the size should not be zero when there is
+    // no columns, this help to prevent divide-by-zero error.
+    val childRowSize = child.output.map(_.dataType.defaultSize).sum + 8
+    val outputRowSize = output.map(_.dataType.defaultSize).sum + 8
+    // Assume there will be the same number of rows as child has.
+    var sizeInBytes = (child.statistics.sizeInBytes * outputRowSize) / childRowSize
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: I see, e.g. if the input is `Long` but output is `Int` then our size should be half our child's. I think it would be slightly clearer if this read
```
child.statistics.sizeInBytes * (outputRowSize / childRowSize)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 2]
```diff
@@ -316,6 +316,21 @@ abstract class UnaryNode extends LogicalPlan {
   override def children: Seq[LogicalPlan] = child :: Nil
 
   override protected def validConstraints: Set[Expression] = child.constraints
+
+  override def statistics: Statistics = {
+    // There should be some overhead in Row object, the size should not be zero when there is
+    // no columns, this help to prevent divide-by-zero error.
+    val childRowSize = child.output.map(_.dataType.defaultSize).sum + 8
+    val outputRowSize = output.map(_.dataType.defaultSize).sum + 8
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: how did you arrive at 8?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala [hunk 3]
```diff
@@ -156,6 +156,13 @@ case class Intersect(left: LogicalPlan, right: LogicalPlan) extends SetOperation
       Some(children.flatMap(_.maxRows).min)
     }
   }
+
+  override def statistics: Statistics = {
+    val leftSize = left.statistics.sizeInBytes
+    val rightSize = right.statistics.sizeInBytes
+    val sizeInBytes = if (leftSize < rightSize) leftSize else rightSize
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: can't you do `math.min`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 4]
```diff
@@ -316,6 +316,21 @@ abstract class UnaryNode extends LogicalPlan {
...
+    // Assume there will be the same number of rows as child has.
+    var sizeInBytes = (child.statistics.sizeInBytes * outputRowSize) / childRowSize
+    if (sizeInBytes == 0) {
+      // sizeInBytes can't be zero, or sizeInBytes of BinaryNode will also be zero
+      // (product of children).
+      sizeInBytes = 1
+    }
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: I wonder if this check should go into the constructor of `Statistics` itself, since we do it in a few places.

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 5]
```diff
@@ -63,36 +63,40 @@ class JoinSuite extends QueryTest with SharedSQLContext {
...
-      ("SELECT * FROM testData full outer join testData2 ON key = a",
-        classOf[SortMergeOuterJoin]),
-      ("SELECT * FROM testData left JOIN testData2 ON (key * a != key + a)",
-        classOf[BroadcastNestedLoopJoin]),
-      ("SELECT * FROM testData right JOIN testData2 ON (key * a != key + a)",
-        classOf[BroadcastNestedLoopJoin]),
-      ("SELECT * FROM testData full JOIN testData2 ON (key * a != key + a)",
-        classOf[BroadcastNestedLoopJoin])
-    ).foreach { case (query, joinClass) => assertJoin(query, joinClass) }
+    withSQLConf("spark.sql.autoBroadcastJoinThreshold" -> "0") {
```
Comment (andrewor14) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: why disable auto broadcast join here? Do all of the following tests become broadcast joins?

**Comment by andrewor14** [comment]
LGTM. The changes are relatively straightforward. The only thing I'm not super sure of is how changing the default values of the various data types will affect things. Maybe others like @nongli or @marmbrus can confirm that it's OK to do so.

**Comment by andrewor14** [comment]
Hm it seems to always fail the same pyspark test

**Comment by davies** [comment]
@andrewor14 I saw that, will investigate it.

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala (hunk 5 above)
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: Some of them will became broadcast joins, should I update the tests without disabling auto-broadcast?

**Comment by davies** [comment]
@andrewor14 I had update the description to explain more on change of defaultSize

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala (hunk 2 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: Yes. 8 is the size of word on x86-64

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala (hunk 1 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: Because outputRowSize could be smaller than childRowSize, `outputRowSize / childRowSize` will become zero

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala (hunk 4 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: I'd like to do that, but Statistics is an case class, I'd not to change the interface. Is there a good way to do that?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala (hunk 3 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: No, they are BitInt, math.min can't work with them.

**Comment by davies** [comment]
@rxin @nongli  This PR help a lot on performance, could you also take a look?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala [hunk 6]
```diff
@@ -521,9 +540,7 @@ case class Expand(
     AttributeSet(projections.flatten.flatMap(_.references))
 
   override def statistics: Statistics = {
-    // TODO shouldn't we factor in the size of the projection versus the size of the backing child
-    //      row?
-    val sizeInBytes = child.statistics.sizeInBytes * projections.length
+    val sizeInBytes = super.statistics.sizeInBytes * projections.length
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: Is this right? This assumes each projection is the identity right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala [hunk 7]
```diff
@@ -74,7 +74,7 @@ abstract class UserDefinedType[UserType] extends DataType with Serializable {
   /**
    * The default size of a value of the UserDefinedType is 4096 bytes.
    */
-  override def defaultSize: Int = 4096
+  override def defaultSize: Int = sqlType.defaultSize
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala]: Why did you change this one? Also doesn't match comment anymore (I think specifying the value in the comment is redundant and not useful)

**Comment by nongli** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala (hunk 6 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: All the projection should produce the same schema (same type for each column).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala (hunk 7 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala]: The old value is bad guess, because we always materialize the rows as sqlType, so should use that. Will remove the comment while merging.

**Comment by davies** [comment]
Merging this into master, thanks!
